plugin/PlayAssetDelivery/src/main/java/com/google/play/core/godot/assetpacks/PlayAssetDelivery.java

```diff
@@ -95,12 +109,60 @@ public void onMainDestroy() {
   void registerAssetPackStateUpdatedListener() {
     assetPackManager.registerListener(
         state -> {
-          emitSignalWrapper(
-              ASSET_PACK_STATE_UPDATED,
-              PlayAssetDeliveryUtils.convertAssetPackStateToDictionary(state));
+          boolean isTerminalState = assetPackTerminalStates.contains(state.status());
+          if (isTerminalState) {
+            ongoingAssetPackRequests.remove(state.name());
+          } else {
+            ongoingAssetPackRequests.add(state.name());
+          }
+
+          emitNonDuplicateStateUpdatedSignal(state);
         });
   }
 
+  /**
+   * Calls getPackStates on all asset packs currently in non-terminal state and emit non-duplicating
+   * stateUpdated signals.
+   */
+  private void forceAssetPackStateUpdate() {
+    assetPackManager
+        .getPackStates(new ArrayList<>(ongoingAssetPackRequests))
+        .addOnSuccessListener(
+            result ->
+                result
+                    .packStates()
+                    .entrySet()
+                    .stream()
+                    .forEach(
+                        e -> {
+                          String packName = e.getKey();
```

Comment:
This code seems a lot like the code in registerAssetPackStateUpdatedListener. Can we factor this out so forceAssetPackStateUpdate and registerAssetPackStateUpdatedListener reference the same method?

Reply:
Refactored to ```emitNonDuplicateStateUpdatedSignal(AssetPackState assetPackState, boolean addToOngoingAssetPackRequests)```